File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  CHATROOMS = Hash.new { |hash, key| hash[key] = [] }.tap do |rooms|
    100.times do |i|
      rooms["main#{i}"] = [] # Add default rooms
    end
  end

  def subscribed
    # Ensure a room is provided
    if params[:room].present?
      # Create the room if it doesn't exist
      CHATROOMS[params[:room]] ||= []

      # Stream from the specific room
      stream_from "chat_channel_#{params[:room]}"
    else
      reject
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def send_message(data)
    puts "Received message: #{data.inspect}"
    message = data['message']
    room = message['room']
    sendable = {
      name: message['name'],
      id: message['id'],
      message: message['message']
    }

    # Add the message to the specific chatroom
    CHATROOMS[room] << sendable

    # Broadcast the message to the specific chatroom
    ActionCable.server.broadcast("chat_channel_#{room}", sendable)
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "./consumer"

const chat = (room) => consumer.subscriptions.create(
  { channel: "ChatChannel", room: room },
  {
    connected() {
      // Called when the subscription is ready for use on the server
      console.log("Connected to the chat channel");
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received(data) {
      // Called when there's incoming data on the websocket for this channel
      console.log("Received data: ", data);
      $('#dummy_message')
        .clone()
        .text(data['chat'])
        .appendTo($('#messages'));
    },

    send_message: function(id, name, message) {
      console.log("Sending message: " + message);
      return this.perform('send_message', { message: { id: id, name: name, room: roomName, message: message } });
    }
  }
);

window.chat = chat;